src/docs/memory-management.md

```diff
@@ -0,0 +1,138 @@
+store ptr %1, ptr %.refType, align 8, !dbg !156
+call swiftcc void @"output.PassThrough(vtype: output.VType) -> ()"(ptr noalias nocapture dereferenceable(8) %0), !dbg !157
+%3 = call ptr @"outlined destroy of output.VType"(ptr %0), !dbg !158
+```
+
+### Parameters passed by reference (inout)
+
+When `inout` reference type or a value type containing a reference property is passed as a parameter, it is passed by reference, so no counters are updated.
+```
+%vtype.debug = alloca ptr, align 8
+call void @llvm.dbg.declare(metadata ptr %vtype.debug, metadata !285, metadata !DIExpression(DW_OP_deref)), !dbg !286
+call void @llvm.memset.p0.i64(ptr align 8 %vtype.debug, i8 0, i64 8, i1 false)
+store ptr %0, ptr %vtype.debug, align 8, !dbg !287
+ret void, !dbg !288
+```
+
+### Return types
+
+When a value type is returned from a function, the caller takes ownership of the instance and is responsible for releasing it at the end of the block. In the callee:
+ - If a new instance is created, its counter is initialized to 1
+ - If an existing parameter is returned, its reference count is retained before returning
+ - If an inout parameter is returned, a copy is created and returned
+
+Pass-through callee:
+```
+%test.debug = alloca ptr, align 8
+call void @llvm.dbg.declare(metadata ptr %test.debug, metadata !265, metadata !DIExpression()), !dbg !266
+call void @llvm.memset.p0.i64(ptr align 8 %test.debug, i8 0, i64 8, i1 false)
+%.refType = getelementptr inbounds %T6output5VTypeV, ptr %1, i32 0, i32 0, !dbg !267
+%2 = load ptr, ptr %.refType, align 8, !dbg !267
+store ptr %2, ptr %test.debug, align 8, !dbg !269
+%3 = call ptr @swift_retain(ptr returned %2) #5, !dbg !270
+%.refType1 = getelementptr inbounds %T6output5VTypeV, ptr %0, i32 0, i32 0, !dbg !270
+store ptr %2, ptr %.refType1, align 8, !dbg !270
+ret void, !dbg !271
+```
+
+New instance callee:
+```
+call swiftcc void @"$s6output5VTypeVACycfC"(ptr noalias nocapture sret(%T6output5VTypeV) %0), !dbg !302
+ret void, !dbg !303
+```
+
+## Memory handling
+
+To handle native memory in the scenarios above, the projections should could use the value witness table to invoke `InitWithCopy` for copy operations and `Destroy` for finalization. These functions manage reference counts at any level of nesting.
+
+To ensure correct memory handling:
+ - Swift value types that contain reference properties should be projected as C# classes
+ - When a type goes out of the block the destructor/dispose should invoke `Destroy` function
+ - When a type is marshalled to Swift as a function parameter, `InitWithCopy` should be invoked to create the copy
+ - When a type is marshalled to Swift as an `inout` function parameters, an instance reference is passed
+ - When a type is marshalled to Swift as a return paramter, `InitWithCopy` should be invoked to create the copy
+ - When a type is marshalled from Swift as a return paramter, a created instance in a callee is consumed and counters are not updated
```

Comment:
```suggestion
 - When a type is marshalled from Swift as a return parameter, a created instance in a callee is consumed and counters are not updated
```

Reply:
> a created instance in a callee is consumed 

Consumed as in, returned to the caller?

"counters" == "ref count"?

Reply:
Simplified with "no reference counters are updated"

> "counters" == "ref count"?

Yes, thanks